kubernetes: No CEP CRD for cilium-health EP

raybejjani opened the pull request:
The cilium-health endpoint doesn't exist in kubernetes land, and we've
had random errors trying to maintain corresponding CEPs. Instead of
playing wack-a-mole with the errors, avoid creating the CEP to begin
with. Older, already created, cilium-health CEPs will be GCed.

Fixes #4979

```release-note/minor
cilium-agent no longer creates a CiliumEndpoint k8s CRD for internal cilium-health endpoints
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4981)
<!-- Reviewable:end -->

raybejjani committed:
kubernetes: No CEP CRD for cilium-health EP

The cilium-health endpoint doesn't exist in kubernetes land, and we've
had random errors trying to maintain corresponding CEPs. Instead of
playing wack-a-mole with the errors, avoid creating the CEP to begin
with. Older, already created, cilium-health CEPs will be GCed.

Signed-off-by: Ray Bejjani <ray@covalent.io>

ianvernon commented on pkg/endpoint/endpoint.go:
```diff
@@ -518,6 +514,12 @@ func (e *Endpoint) RunK8sCiliumEndpointSync() {
 		return
 	}
 
+	if isHealthEP := e.HasLabels(pkgLabels.LabelHealth); isHealthEP {
```
This will deadlock the agent.

The caller of `RunK8sCiliumEndpointSync` holds the endpoint's mutex:
https://github.com/cilium/cilium/blob/9639c30f14ea7428d182452a13a0d7ad6a5fbc24/pkg/endpointmanager/manager.go#L61-L70

And `endpoint.HasLabels` tries to `RLock()` that same lock:
https://github.com/cilium/cilium/blob/9639c30f14ea7428d182452a13a0d7ad6a5fbc24/pkg/endpoint/endpoint.go#L1660-L1681

ianvernon replied:
Oops, the mutex I am looking at is the endpointmanager mutex in `Insert`...

raybejjani replied:
but! there is a deadlock detected on `ep.Mutex`, so you have magic intuition.

ianvernon replied:
https://stackoverflow.com/questions/30547916/goroutine-blocks-when-calling-rwmutex-rlock-twice-after-an-rwmutex-unlock `Insert` for an endpoint requires that the endpoint's mutex is read-locked. `HasLabels` calls `RLock` on the same mutex, within the same goroutine. However, this will cause deadlock in the code, because of the documentation for Golang's `RWMutex`:

> To ensure that the lock eventually becomes available, a blocked Lock call excludes new readers from acquiring the lock.

This effectively means that trying `RLock` twice within the same goroutine can result in a deadlock. 

Source: https://stackoverflow.com/questions/30547916/goroutine-blocks-when-calling-rwmutex-rlock-twice-after-an-rwmutex-unlock

ianvernon reviewed (changes requested):
Please address the comment re: deadlock.

ianvernon commented on pkg/endpoint/endpoint.go:
```diff
@@ -535,29 +531,30 @@ func (e *Endpoint) RunK8sCiliumEndpointSync() {
+				// RunK8sCiliumEndpointSync hold ep.Mutex and that conflicts with
+				// HasLabels.
+				if isHealthEP := e.HasLabels(pkgLabels.LabelHealth); isHealthEP {
+					// use an error object so we can stop the controller. This is NOT an
```
Instead of starting the controller, why not create a private function, `hasLabelsRLocked` which assumes the RLock is held and then call it outside of the controller? You can use this function within `HasLabels`. This solves the potential deadlock. We are trying to induce behavior that indicates the controller shouldn't be ran; we should not run the controller and add the necessary code to avoid deadlock.

raybejjani replied:
I had hoped to keep the controller create function free of the endpoint. At least, not have it hold a lock. As it turns out, we create this controller on `e.controllers` (it was once unattached to a local controller, only the global one).
With that in mind, what you're suggesting works far better. I've also labelled the function as needing the rlock.

ianvernon commented on pkg/controller/controller.go:
```diff
@@ -32,13 +33,19 @@ type ControllerFunc func() error
 
+// MakeExitReason constructs an error object that, when returned by a DoFunc,
+// stops the controller from running further.
+func MakeExitReason(msg string) ExitReason {
```
Unit test this please, and plumb its usage wherever creation of an exit reason is called directly in the controller code.

ianvernon reviewed (changes requested):
See comments.

raybejjani committed:
pkg/endpoint: Factor out Endpoint.HasLabelsRLocked

We need to check the labels of the endpoint while holding locks. This
wasn't exposed and now is.

Signed-off-by: Ray Bejjani <ray@covalent.io>

raybejjani reviewed (commented):
Ok, this should be less wonky now.
